quic/congestion_control/Copa2.cpp

```diff
@@ -0,0 +1,262 @@
+  if (loss.persistentCongestion) {
+    VLOG(10) << __func__ << " writable=" << getWritableBytes()
+             << " cwnd=" << cwndBytes_
+             << " inflight=" << conn_.lossState.inflightBytes << " " << conn_;
+    cwndBytes_ = conn_.transportSettings.minCwndInMss * conn_.udpSendPacketLen;
+    if (conn_.pacer) {
+      // TODO Which min RTT should we use?
+      conn_.pacer->refreshPacingRate(cwndBytes_, conn_.lossState.mrtt);
+    }
+    if (conn_.qLogger) {
+      conn_.qLogger->addCongestionMetricUpdate(
+          conn_.lossState.inflightBytes,
+          getCongestionWindow(),
+          kPersistentCongestion);
+    }
+  }
+
+  numLostInLossCycle_ += loss.lostPackets;
+  manageLossyMode(loss.largestLostSentTime);
+}
+
+void Copa2::onPacketAcked(const AckEvent& ack) {
+  DCHECK(ack.largestAckedPacket.has_value());
+  subtractAndCheckUnderflow(conn_.lossState.inflightBytes, ack.ackedBytes);
+  minRTTFilter_.Update(
+      conn_.lossState.lrtt,
+      std::chrono::duration_cast<microseconds>(ack.ackTime.time_since_epoch())
+          .count());
+  bytesAckedInCycle_ += ack.ackedBytes;
+  for (const auto& ackPkt : ack.ackedPackets) {
+    appLimitedInCycle_ = appLimitedInCycle_ || ackPkt.isAppLimited;
+  }
+  auto rttMin = minRTTFilter_.GetBest();
+
+  numAckedInLossCycle_ += ack.ackedPackets.size();
+  manageLossyMode(ack.largestAckedPacketSentTime);
+
+  auto dParam = rttMin;
+  if (lossyMode_) {
+    dParam = duration_cast<microseconds>(rttMin * 2. * lossToleranceParam_);
+  }
+  auto cycleDur = rttMin + dParam;
+
+  if (probeRtt_) {
+    // Do we exit probe RTT now?
+    if (lastProbeRtt_ + dParam <= ack.ackTime) {
+      // Note, probe rtt should ideally never decrease ack rate, since
+      // it just barely empties the queue. Hence all ack rate samples
+      // are good independent of whether we probed for rtt
+      probeRtt_ = false;
+    }
+  } else {
+    // See if we need to enter probe rtt mode
+    if (conn_.lossState.lrtt < rttMin + dParam) {
+      // RTT has fallen below threshold. Maybe others are probing. As long as we
+      // are not doing this too often, we should enter probe rtt too
+      if (lastProbeRtt_ + kCopa2ProbeRttInterval / 2 <= ack.ackTime) {
+        probeRtt_ = true;
+        lastProbeRtt_ = ack.ackTime;
+      }
+    } else {
+      // If it has been long enough since the last time we entered probeRtt,
+      // enter again
+      if (lastProbeRtt_ + kCopa2ProbeRttInterval <= ack.ackTime) {
+        probeRtt_ = true;
+        lastProbeRtt_ = ack.ackTime;
+      }
+    }
+  }
+
+  if (!cycleStartTime_) {
+    cycleStartTime_ = ack.ackTime;
+    return;
+  }
+
+  // See if cycle needs to be continued
+  if (*cycleStartTime_ + cycleDur > ack.ackTime) {
+    return;
+  }
+
+  // Cycle has ended. Update cwnd and rate
+  auto newCwnd = bytesAckedInCycle_ + alphaParam_ * conn_.udpSendPacketLen;
+  if (!appLimitedInCycle_ || cwndBytes_ < newCwnd) {
```

Comment:
i think given appLimitedInCycle_ will always be false, then this is equivalent to never decrease cwnd.